chore: add faqs to README.md (#7)
chadwatson/react-remote-resource: README.md

  </div>
);
```

&nbsp;

---

&nbsp;

> ## FAQ

Reviewer: I think we could also stand to add a different section talking about how RRR doesn't care what shape your data is in. Maybe something about how whatever your Promise resolves with is what we store. What do you think?

I'm thinking that people may not immediately realize that they can just chain `.then`s forever and whatever they end up with will be used.

Author: Hrm sounds like two questions:

> Does my data need to be in a specific shape to use `react-remote-resource`?

> What are entries and how are entries created?

Reviewer: I think that sounds right.

> ### When should I use `createKeyedResource` vs `createSingleEntryResource`?

The the main difference between `createKeyedResource` and `createSingleEntryResource` is how the created `resource` stores its entries. Internally, every `resource` organizes the data it receives from the completed load function into entries. A `resource`'s main concern is how the data in the entries should be available to your app, not necessarily how the external api is structured or called.

- `createKeyedResource` creates a resource with multiple entries that are organized by key. [See the first parameter on how entries are keyed](https://github.com/chadwatson/react-remote-resource#createkeyedresource).

- `createSingleEntryResource` creates a resource with only a single entry. All data, regardless of the structure, will be stored in only one entry.

> Here are some examples scenerios that show when each of the :

Reviewer: Looks like this is missing some stuff.

Author: On it!

#### Example Users API:

Assuming You have a users api that takes an id and returns user information:

```javascript
/*
  
  /api/users/:id

  Example Response:
  {
    name: "name",
    id: {userId},
    ...
  }

*/
```

**Scenerio 1:**

> The app only needs the current users information. Since the app only one entry from the api (the current users information and no others), `createSingleEntryResource` would work well.

Reviewer: ```suggestion
The app only needs the current user's information. Since the app only needs one entry from the api (the current user's information and no others), `createSingleEntryResource` would work well.
```

```jsx

> const load = (id) => fetch(`/api/users/${id}`);

Reviewer: ```suggestion
const load = id => fetch(`/api/users/${id}`);
```

const userResource = createSingleEntryResource(load);

const currentUserId = ...;

const AboutMe = () => {
  const [user] = userResource.useEntry(currentUserId);

  return ...;

> }

Reviewer: Might be clearer to just supply a hard-coded ID to `useEntry`.

```jsx
const load = (id) => fetch(`/api/users/${id}`);
const userResource = createSingleEntryResource(load);

const AboutMe = () => {
  const [user] = userResource.useEntry(12345);
  return ...;
};
```

Author: Yes it is!

```

**Scenerio 2**

> The app needs multiple users from the API. Since the app needs multiple entries from the api organized by key then createKeyedResource would then be a better choice.

Reviewer: ```suggestion
The app needs to make individual requests to the same users API and store the results independently. `createKeyedResource` would be the better choice.
```

```jsx

> const load = (id) => fetch(`/api/users/${id}`);

Reviewer: ```suggestion
const load = id => fetch(`/api/users/${id}`);
```

const usersResource = createKeyedResource(id => id, load);

> const User = ({id}) => {

Reviewer: ```suggestion
const User = ({ id }) => {
```

  const [user] = usersResource.useEntry(id);

  return ...;
}

> const UserList = ({ids}) => {

Reviewer: ```suggestion
const UserList = ({ ids }) => {
```

>   return ids.map(id => <User key={id} id={id} />

Reviewer: ```suggestion
  return ids.map(id => <User key={id} id={id} />);
```

}
```

### Example Clients API:

> Assuming You have a clients api that takes an account_rep_id and returns a list of clients for that specific account rep:

Reviewer: ```suggestion
Assuming you have a clients API that takes an `account_rep_id` and returns a list of clients for that specific account rep:
```

```javascript
/*

/api/clients/:account_rep_id

{
  123454: {
      ...
  },
  508923: {
      ...
  },
  14: {
      ...
  },
  995: {
      ...
  }

> }

Reviewer: Since the description above says the API returns a "list of clients" should the data be more shaped like this?

```javascript
[
  {
    id: 123454,
    ...
  },
  {
    id: 508923,
    ...
  },
  {
    id: 14,
    ...
  },
  {
    id: 995,
    ...
  }
]
```

Author: Sounds good!

*/
```

**Scenerio 1:**

> The app only needs the current account rep's list: createSingleEntryResource would work well.

Reviewer: ```suggestion
The app only needs the current account rep's list: `createSingleEntryResource` would work well.
```

```jsx
const load = (id) => fetch(`/api/clients/${id}`);

const clientsResource = createSingleEntryResource(load);

const accountRepId = ...;

const ClientList = () => {
  const [clients] = clientsResource.useEntry(accountRepId);

  return ...;
}

> ```

Reviewer: Similar comment to one of the examples above:

```jsx
const load = (id) => fetch(`/api/clients/${id}`);
const clientsResource = createSingleEntryResource(load);

const ClientList = () => {
  const [clients] = clientsResource.useEntry(12345);
  return ...;
};
```

Author: Kk!

**Scenerio 2:**

> The app needs multiple account rep's client list: This is suited better for createKeyedResource.

Reviewer: ```suggestion
The app needs multiple account rep's client list: `createKeyedResource` is better suited.
```

```jsx

> const load = (account_rep_id) => fetch(`/api/clients/${account_rep_id}`);

Reviewer: ```suggestion
const load = account_rep_id => fetch(`/api/clients/${account_rep_id}`);
```

const clientsResource = createKeyedResource(id => id, load);

> const ClientList = ({account_rep_id}) => {

Reviewer: ```suggestion
const ClientList = ({ account_rep_id }) => {
```

  const [clients] = clientsResource.useEntry(account_rep_id);

  return ...;
}
```

### Example Posts API:

You have a posts api that takes no parameters and returns a list of posts organized by post id:

```javascript
/*

/api/posts

{
  882: {
    userId: 5,
    ...
  },
  622: {
    userId: 10,
    ...
  },
  1: {
    userId: 1,
    ...
  },
  102: {
    userId: 10,
    ...
  }

> }

Reviewer: Similar to a previous comment:

```javascript
[
  {
    id: 882,
    userId: 5
  },
  {
    id: 622,
    userId: 10
  },
  {
    id: 622,
    userId: 10
  },
  {
    id: 1,
    userId: 1
  },
  {
    id: 102,
    userId: 10
  }
]
```

Author: Sounds good!

*/
```

The app itself needs the data to be organized by userId.

`createSingleEntryResource` could be utilized and composed for this:

```jsx
const postsResource = createSingleEntryResource(() =>
  fetch("/api/posts").then(posts =>
    posts.reduce(
      (acc, post) => ({
        ...acc,
        [post.userId]: acc[post.userId] ? [...acc[post.userId], post] : [post]
      }),
      {}
    )
  )
);

const useUsersPosts = id => {
  const [allPosts] = postsResources.useEntry();
  return allPosts[id] || [];
};
```

### Last Note

> `createSingleEntryResource` and `createKeyedResource` are composed versions of `createResource`. If they do not fit your use case, try using `createResource` directly, as it may better suit your needs.

Reviewer: Good word!

## FAQ

> ### What are resource entries and how are they created?

An entry is data that is stored in the `resource` when the promise of the `load` finally resolves.

Each [Resource Creator](https://github.com/chadwatson/react-remote-resource#resource-creators) determines how the entries are organized and stored. As an example, `createSingleEntryResource` creates a `resource` that only has a single entry from an api.

```javascript
const postsResource = createSingleEntryResource(() =>
  fetch("/api/posts")
    .then(normalizePosts)
    .then(filterToCurrentUser)
    .then(...)
);
```

The `postsResource` above, will store the value as a single entry in the resource once all the `.then` statements complete.

&nbsp;

> ### Does my data need to be in a specific shape to use `react-remote-resource`?

> No! `react-remote-resource` aims to h Whatever value resolves from the promise of the `load`, regardless of shape or type, will be stored as an entry.

Reviewer: Little typo here.

Author: I (missing a bunch here). That should do it... 😂 Haha

Reviewer: Great update.
